File: Backend/controllers/uploadController.js
import User from "../models/User.js";
import Provider from "../models/Provider.js";
import Service from "../models/Service.js";
import { deleteImage } from "../config/cloudinary.js";
import { apiResponse } from "../utils/helpers.js";

// Upload user profile image
export const uploadUserProfileImage = async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json(apiResponse(false, "No image file provided"));
    }

    const user = await User.findById(req.userId);
    if (!user) {
      return res.status(404).json(apiResponse(false, "User not found"));
    }

    // Delete old image if exists
    if (user.profileImage) {
      await deleteImage(user.profileImage);
    }

    // Update user with new image URL
    user.profileImage = req.file.path;
    user.updatedAt = Date.now();
    await user.save();

    res.json(
      apiResponse(true, "Profile image uploaded successfully", {
        profileImage: user.profileImage,
      }),
    );
  } catch (error) {
    console.error("Error uploading user profile image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to upload image", error.message));
  }
};

// Delete user profile image
export const deleteUserProfileImage = async (req, res) => {
  try {
    const user = await User.findById(req.userId);
    if (!user) {
      return res.status(404).json(apiResponse(false, "User not found"));
    }

    if (user.profileImage) {
      await deleteImage(user.profileImage);
      user.profileImage = null;
      user.updatedAt = Date.now();
      await user.save();
    }

    res.json(apiResponse(true, "Profile image deleted successfully"));
  } catch (error) {
    console.error("Error deleting user profile image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to delete image", error.message));
  }
};

// Upload provider profile image
export const uploadProviderProfileImage = async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json(apiResponse(false, "No image file provided"));
    }

    const provider = await Provider.findOne({ userId: req.userId });
    if (!provider) {
      return res.status(404).json(apiResponse(false, "Provider not found"));
    }

    // Delete old image if exists
    if (provider.profileImage) {
      await deleteImage(provider.profileImage);
    }

    // Update provider with new image URL
    provider.profileImage = req.file.path;
    provider.updatedAt = Date.now();
    await provider.save();

    res.json(
      apiResponse(true, "Provider profile image uploaded successfully", {
        profileImage: provider.profileImage,
      }),
    );
  } catch (error) {
    console.error("Error uploading provider profile image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to upload image", error.message));
  }
};

// Delete provider profile image
export const deleteProviderProfileImage = async (req, res) => {
  try {
    const provider = await Provider.findOne({ userId: req.userId });
    if (!provider) {
      return res.status(404).json(apiResponse(false, "Provider not found"));
    }

    if (provider.profileImage) {
      await deleteImage(provider.profileImage);
      provider.profileImage = null;
      provider.updatedAt = Date.now();
      await provider.save();
    }

    res.json(apiResponse(true, "Provider profile image deleted successfully"));
  } catch (error) {
    console.error("Error deleting provider profile image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to delete image", error.message));
  }
};

// Upload service image
export const uploadServiceImage = async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json(apiResponse(false, "No image file provided"));
    }

    const { serviceId } = req.params;

    // Get provider for current user
    const provider = await Provider.findOne({ userId: req.userId });
    if (!provider) {
      return res.status(404).json(apiResponse(false, "Provider not found"));
    }

    // Find service and verify ownership
    const service = await Service.findById(serviceId);
    if (!service) {
      return res.status(404).json(apiResponse(false, "Service not found"));
    }

    if (service.providerId.toString() !== provider._id.toString()) {
      return res
        .status(403)
        .json(apiResponse(false, "Not authorized to update this service"));
    }

    // Delete old image if exists
    if (service.image) {
      await deleteImage(service.image);
    }

    // Update service with new image URL
    service.image = req.file.path;
    service.updatedAt = Date.now();
    await service.save();

    res.json(
      apiResponse(true, "Service image uploaded successfully", {
        image: service.image,
      }),
    );
  } catch (error) {
    console.error("Error uploading service image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to upload image", error.message));
  }
};

// Delete service image
export const deleteServiceImage = async (req, res) => {
  try {
    const { serviceId } = req.params;

    // Get provider for current user
    const provider = await Provider.findOne({ userId: req.userId });
    if (!provider) {
      return res.status(404).json(apiResponse(false, "Provider not found"));
    }

    // Find service and verify ownership
    const service = await Service.findById(serviceId);
    if (!service) {
      return res.status(404).json(apiResponse(false, "Service not found"));
    }

    if (service.providerId.toString() !== provider._id.toString()) {
      return res
        .status(403)
        .json(apiResponse(false, "Not authorized to update this service"));
    }

    if (service.image) {
      await deleteImage(service.image);
      service.image = null;
      service.updatedAt = Date.now();
      await service.save();
    }

    res.json(apiResponse(true, "Service image deleted successfully"));
  } catch (error) {
    console.error("Error deleting service image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to delete image", error.message));
  }
};

// Admin: Upload service image (for any service)
export const adminUploadServiceImage = async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json(apiResponse(false, "No image file provided"));
    }

    const { serviceId } = req.params;

    const service = await Service.findById(serviceId);
    if (!service) {
      return res.status(404).json(apiResponse(false, "Service not found"));
    }

    // Delete old image if exists
    if (service.image) {
      await deleteImage(service.image);
    }

    // Update service with new image URL
    service.image = req.file.path;
    service.updatedAt = Date.now();
    await service.save();

    res.json(
      apiResponse(true, "Service image uploaded successfully", {
        image: service.image,
      }),
    );
  } catch (error) {
    console.error("Error uploading service image:", error);
    res
      .status(500)
      .json(apiResponse(false, "Failed to upload image", error.message));
  }
};

File: Backend-SpringBoot/src/main/java/com/servesync/controller/UploadController.java
package com.servesync.controller;

import com.servesync.dto.ApiResponse;
import com.servesync.model.Provider;
import com.servesync.model.User;
import com.servesync.repository.ProviderRepository;
import com.servesync.repository.UserRepository;
import com.servesync.service.UploadService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

@RestController
@RequestMapping("/api/upload")
public class UploadController {
    
    @Autowired
    private UploadService uploadService;
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private ProviderRepository providerRepository;
    
    @PostMapping("/image")
    public ResponseEntity<ApiResponse<Map<String, String>>> uploadImage(
            @RequestParam("image") MultipartFile image,
            @RequestParam(value = "folder", defaultValue = "servesync") String folder) {
        try {
            String imageUrl = uploadService.uploadImage(image, folder);
            return ResponseEntity.ok(ApiResponse.success("Image uploaded successfully", 
                    Map.of("url", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @PostMapping("/user/profile-image")
    public ResponseEntity<ApiResponse<Map<String, String>>> uploadUserProfileImage(
            @RequestParam("image") MultipartFile image,
            @AuthenticationPrincipal User user) {
        try {
            String imageUrl = uploadService.uploadImage(image, "servesync/profiles");
            
            user.setProfileImage(imageUrl);
            userRepository.save(user);
            
            return ResponseEntity.ok(ApiResponse.success("Profile image uploaded successfully", 
                    Map.of("profileImage", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @DeleteMapping("/user/profile-image")
    public ResponseEntity<ApiResponse<String>> deleteUserProfileImage(
            @AuthenticationPrincipal User user) {
        try {
            if (user.getProfileImage() != null) {
                uploadService.deleteImage(user.getProfileImage());
                user.setProfileImage(null);
                userRepository.save(user);
            }
            return ResponseEntity.ok(ApiResponse.success("Profile image deleted successfully"));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @PostMapping("/provider/profile-image")
    public ResponseEntity<ApiResponse<Map<String, String>>> uploadProviderProfileImage(
            @RequestParam("image") MultipartFile image,
            @AuthenticationPrincipal User user) {
        try {
            // Find provider by userId
            Provider provider = providerRepository.findByUserId(user.getId())
                    .orElseThrow(() -> new RuntimeException("Provider not found"));
            
            String imageUrl = uploadService.uploadImage(image, "servesync/providers");
            
            provider.setProfileImage(imageUrl);
            providerRepository.save(provider);
            
            return ResponseEntity.ok(ApiResponse.success("Provider profile image uploaded successfully", 
                    Map.of("profileImage", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @DeleteMapping("/provider/profile-image")
    public ResponseEntity<ApiResponse<String>> deleteProviderProfileImage(
            @AuthenticationPrincipal User user) {
        try {
            // Find provider by userId
            Provider provider = providerRepository.findByUserId(user.getId())
                    .orElseThrow(() -> new RuntimeException("Provider not found"));
            
            if (provider.getProfileImage() != null) {
                uploadService.deleteImage(provider.getProfileImage());
                provider.setProfileImage(null);
                providerRepository.save(provider);
            }
            return ResponseEntity.ok(ApiResponse.success("Provider profile image deleted successfully"));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @PostMapping("/service/{serviceId}/image")
    public ResponseEntity<ApiResponse<Map<String, String>>> uploadServiceImageWithId(
            @PathVariable String serviceId,
            @RequestParam("image") MultipartFile image) {
        try {
            String imageUrl = uploadService.uploadImage(image, "servesync/services");
            return ResponseEntity.ok(ApiResponse.success("Service image uploaded successfully", 
                    Map.of("url", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @DeleteMapping("/service/{serviceId}/image")
    public ResponseEntity<ApiResponse<String>> deleteServiceImageWithId(
            @PathVariable String serviceId) {
        try {
            return ResponseEntity.ok(ApiResponse.success("Service image deleted successfully"));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @PostMapping("/admin/service/{serviceId}/image")
    public ResponseEntity<ApiResponse<Map<String, String>>> adminUploadServiceImage(
            @PathVariable String serviceId,
            @RequestParam("image") MultipartFile image) {
        try {
            String imageUrl = uploadService.uploadImage(image, "servesync/services");
            return ResponseEntity.ok(ApiResponse.success("Service image uploaded successfully", 
                    Map.of("url", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @PostMapping("/service/image")
    public ResponseEntity<ApiResponse<Map<String, String>>> uploadServiceImage(
            @RequestParam("image") MultipartFile image) {
        try {
            String imageUrl = uploadService.uploadImage(image, "servesync/services");
            return ResponseEntity.ok(ApiResponse.success("Service image uploaded successfully", 
                    Map.of("url", imageUrl)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
    
    @DeleteMapping("/image")
    public ResponseEntity<ApiResponse<String>> deleteImage(@RequestBody Map<String, String> body) {
        try {
            String imageUrl = body.get("imageUrl");
            uploadService.deleteImage(imageUrl);
            return ResponseEntity.ok(ApiResponse.success("Image deleted successfully"));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }
}
